crates/hyperdrive-math/src/lp.rs

```diff
@@ -4,9 +4,144 @@ use ethers::types::{I256, U256};
+            } else {
+                I256::try_from(contribution).unwrap()
+            }
+        };
+        let new_state = self.update_liquidity(share_contribution);
+        let ending_present_value = new_state.calculate_present_value(current_block_timestamp);
+
+        // Ensure the present value didn't decreased after adding liquidity.
+        if ending_present_value < starting_present_value {
+            return Err("DecreasedPresentValueWhenAddingLiquidity");
+        }
+
+        // Calculate the lp_shares.
+        let lp_shares = (ending_present_value - starting_present_value)
+            .mul_div_down(lp_total_supply, starting_present_value);
+
+        // Ensure that enough lp_shares are minted so that they can be redeemed.
+        if lp_shares < self.minimum_transaction_amount() {
+            return Err("MinimumTransactionAmount");
+        }
+
+        return Ok(lp_shares);
+    }
+
+    // Gets the resulting state when updating liquidity.
+    pub fn update_liquidity(&self, share_reserves_delta: I256) -> State {
+        let share_reserves = self.share_reserves();
+        let share_adjustment = self.share_adjustment();
+        let bond_reserves = self.bond_reserves();
+        let minimum_share_reserves = self.minimum_share_reserves();
+
+        // Calculate new reserve and adjustment levels.
+        let (updated_share_reserves, updated_share_adjustment, updated_bond_reserves) = self
+            .calculate_update_liquidity(
+                share_reserves,
+                share_adjustment,
+                bond_reserves,
+                minimum_share_reserves,
+                share_reserves_delta,
+            )
+            .unwrap();
+
+        // Update and return the new state.
+        let mut new_info = self.info.clone();
+        new_info.share_reserves = U256::from(updated_share_reserves);
+        new_info.share_adjustment = updated_share_adjustment;
+        new_info.bond_reserves = U256::from(updated_bond_reserves);
+        State {
+            config: self.config.clone(),
+            info: new_info,
+        }
+    }
+
+    // Calculates the resulting share_reserves, bond_reserves and
+    // share_adjustment when updating liquidity with a share_reserves_delta.
+    fn calculate_update_liquidity(
+        &self,
+        share_reserves: FixedPoint,
+        share_adjustment: I256,
+        bond_reserves: FixedPoint,
+        minimum_share_reserves: FixedPoint,
+        share_reserves_delta: I256,
+    ) -> Result<(FixedPoint, I256, FixedPoint), &'static str> {
+        if share_reserves_delta == I256::zero() {
+            return Ok((share_reserves, share_adjustment, bond_reserves));
+        }
+
+        // Get the updated share reserves.
+        let new_share_reserves = if share_reserves_delta > I256::zero() {
+            share_reserves + FixedPoint::from(share_reserves_delta)
+        } else {
+            share_reserves - FixedPoint::from(share_reserves_delta)
```

Comment:
I think we want a check in here that `share_reserves > share_reserves_delta`, otherwise we can get a FixedPoint negative number error. You could combine this with the check on line 117:

```rs
if share_reserves > FixedPoint::from(share_reserves_delta) + minimum_share_reserves {
  share_reserves - FixedPoint::from(share_reserves_delta)
} else {
  return Err("Update would result in share reserves below minimum.");
}
```